contracts/peg/BasePSM.sol

```diff
@@ -0,0 +1,210 @@
+    constructor(OracleParams memory params, IERC20 _underlyingToken)
+        OracleRef(
+            params.coreAddress,
+            params.oracleAddress,
+            params.backupOracle,
+            params.decimalsNormalizer,
+            params.doInvert
+        )
+    {
+        underlyingToken = _underlyingToken;
+    }
+
+    /// @notice withdraw assets from PSM to an external address
+    function withdraw(address to, uint256 amount)
+        external
+        override
+        onlyPCVController
+    {
+        _withdrawERC20(address(underlyingToken), to, amount);
+    }
+
+    /// @notice function to redeem VOLT for an underlying asset
+    function redeem(
+        address to,
+        uint256 amountVoltIn,
+        uint256 minAmountOut
+    ) external virtual override whenNotPaused returns (uint256 amountOut) {
+        amountOut = getRedeemAmountOut(amountVoltIn);
+        require(amountOut >= minAmountOut, "BasePSM: Redeem not enough out");
+
+        _beforeVoltRedeem(to, amountVoltIn, minAmountOut);
+
+        IERC20(volt()).safeTransferFrom(
+            msg.sender,
+            address(this),
+            amountVoltIn
+        );
+
+        underlyingToken.safeTransfer(to, amountOut);
+
+        emit Redeem(to, amountVoltIn, amountOut);
+
+        _afterVoltRedeem(to, amountVoltIn, minAmountOut);
+    }
+
+    /// @notice function to buy VOLT for an underlying asset
+    function mint(
+        address to,
+        uint256 amountIn,
+        uint256 minAmountVoltOut
+    ) external virtual override whenNotPaused returns (uint256 amountVoltOut) {
+        amountVoltOut = getMintAmountOut(amountIn);
+
+        require(
+            amountVoltOut >= minAmountVoltOut,
+            "BasePSM: Mint not enough out"
+        );
+
+        _beforeVoltMint(to, amountIn, minAmountVoltOut);
+
+        underlyingToken.safeTransferFrom(msg.sender, address(this), amountIn);
+
+        IERC20(volt()).safeTransfer(to, amountVoltOut);
+
+        emit Mint(to, amountIn, amountVoltOut);
+
+        _afterVoltMint(to, amountIn, minAmountVoltOut);
+    }
+
+    // ----------- Public State Changing API ----------
+
+    /// @notice function to receive ERC20 tokens from external contracts
+
+    function deposit() external virtual override {}
+
+    // ----------- Public View-Only API ----------
+
+    /// @notice calculate the amount of VOLT out for a given `amountIn` of underlying
+    /// First get oracle price of token
+    /// Then figure out how many dollars that amount in is worth by multiplying price * amount.
+    /// ensure decimals are normalized if on underlying they are not 18
+    function getMintAmountOut(uint256 amountIn)
+        public
+        view
+        virtual
+        override
+        returns (uint256 amountVoltOut)
+    {
+        Decimal.D256 memory oraclePrice = readOracle();
+        _validatePriceRange(oraclePrice);
+
+        amountVoltOut = (amountIn * 1e18) / oraclePrice.value;
```

Comment:
add explanation here as to why this was chosen, with an example of inputs and outputs